lib/pure/views.nim

```diff
@@ -0,0 +1,149 @@
+  let copied = new(seq[T])
+  copied[] = s
+  result = initView(copied)
+
+proc newView*[T](typ: typedesc[T], len: int): View[T] =
+  ## Create uninitialized view of length ``len``.
+  let s = new(seq[T])
+  s[] = newSeq[T](len)
+  result = initView(s)
+
+proc isNil*(v: View): bool =
+  return v.len == 0
+
+proc len*(v: View): int =
+  return v.length
+
+proc ptrAdd[T](p: pointer, i: int): ptr T =
+  return cast[ptr T](cast[int](p) +% (i * sizeof(T)))
+
+proc `[]`*[T](v: View[T], i: int): var T =
+  doAssert(i >= 0 and i < v.len)
+  return ptrAdd[T](v.data, i)[]
+
+proc `[]=`*[T](v: View[T], i: int, val: T) =
+  doAssert(i >= 0 and i < v.len)
+  ptrAdd[T](v.data, i)[] = val
+
+proc slice*[T](v: View[T], start: int, len: int): View[T] =
+  ## Returns a subview starting at ``v[start]`` with length ``len``.
+  if len != 0:
+    doAssert(start < v.len and start >= 0)
+    doAssert(len <= v.len)
+    doAssert(start + len <= v.len)
+    doAssert(len >= 0)
+    result.data = ptrAdd[T](v.data, start)
+    result.length = len
+    when needGcKeep:
+      result.gcKeep = v.gcKeep
+  else:
+    result.data = nil
+    result.length = 0
+
+proc slice*[T](v: View[T], start: int): View[T] =
+  ## Returns a subview starting at ``v[start]`` with length ``v.len - start``.
+  assert start <= v.len and start >= 0
+  return v.slice(start, v.len - start)
+
+# Types that may be safely copied using ``copyMem``.
+type ScalarType = uint8 | uint16 | uint32 | uint64 | int8 | int16 | int32 | int64 | float32 | float64 | byte | char | enum
+
+proc copyFrom*[T](dst: View[T], src: View[T]) =
+  ## Copies content of ``src`` into ``dst``. ``dst.len`` needs to be larger than ``src.len``.
+  doAssert(dst.len >= src.len)
+  when T is ScalarType:
+    copyMem(dst.data, src.data, src.len * sizeof(T))
+  else:
+    for i in 0..<src.len:
+      ptrAdd[T](dst.data, i)[] = ptrAdd[T](src.data, i)[]
+
+proc copyTo*[T](src: View[T], dst: View[T]) =
+  ## Copies content of ``src`` into ``dst``. ``dst.len`` needs to be larger than ``src.len``.
+  dst.copyFrom(src)
+
+proc copyAsSeq*[T](src: View[T]): seq[T] =
+  ## Copies content of ``src`` into a new sequence and returns it.
+  result = newSeq[T](src.len)
+  src.copyTo(initView(addr result[0], result.len))
+
+iterator items*[T](src: View[T]): T =
+  ## Iterate over the content of ``src``.
+  for i in 0..<src.len:
+    yield src[i]
+
+proc `$`*[T](v: View[T]): string =
+  return "View[$1, $2]" % [$v.len, $v.copyAsSeq]
+
+# ByteView
+
+converter toByteView*(s: View[char]): ByteView =
+  # View[char] and View[byte] are mostly the same thing
+  return initView(cast[ptr byte](s.data), s.len, when needGcKeep: s.gcKeep else: nil)
+
+proc initView(s: ref string): ByteView =
+  ## Returns a view into a string.
+  if s[].len == 0:
+    initEmptyView(byte)
+  else:
+    initView(addr s[0], s[].len, gcKeep=s)
+
+proc newView*(s: string): ByteView =
+  ## Copies a string and returns a new view pointing into the copy.
+  let copied = new(string)
```

Comment:
I'm not sure. `shallowCopy` is counter-intuitive here. From `system` docs:
> Be careful with the changed semantics though! There is a reason why the default assignment does a deep copy of sequences and strings.

(notice that `ref string` variant is not exported, because it's not safe when the string is resized)